src/main/java/com/digitalasset/testing/ledger/DamlScriptRunner.java

```diff
@@ -76,19 +83,32 @@ public Builder damlRoot(Path damlRoot) {
       return this;
     }
 
-    public DamlScriptRunner build() {
+    public Builder parties(String[] parties) {
+      this.parties = parties;
+      return this;
```

Comment:
Some assert here would be very developer friendly to ensure that these party names weren't used before.
We can implement in a separate PR, let's not block this. Made a note on the jira.

Reply:
One can only check whether parties contains a duplicate or not - is that enough?

Reply:
I think we need more. We need a mechanism to raise error when the developer forgot to call the obfuscator and simply passes "Bob" in multiple tests:
```
testA()
...
parties(["Bob"]);

testA()
...
parties(["Bob"]); // raises error
```

Reply:
If we want to force the developer to use this feature then we have to do the "obfuscation" underneath, in ledger()/sandbox.
Currently, it's usage is optional and needs to be documented as a best practice, I don't think we have ways to raise error as you write above.

@tamaskalcza-da please weigh in.

Reply:
we have ways, e.g. a `static Set usedPartyNames`.
I'm not sure if we have a fool-proof way to hide this underneath: we'll never know for sure if a party name is "obfuscated" or not. E.g. this is obfuscated:
`party = getContractFromLedgerAndExtractParty(); submitCommand(party);`
but this may not be:
`party = /* handcrafted unique party name, hard to enforce */ ; submitCommand(party);`

Reply:
If we hide it underneath, any function getting a contract should return the "non-obfuscated" party names and all the methods of the API should accept "non-obfuscated" party names as parameters as well.

Reply:
According to Tomi's explanation how JUnit and similar frameworks work, you cannot easily use `static Set usedPartyNames` to save the list of already used party names.
On the other hand, I still think it is a better idea to ask the user to use our `getUniqueParty` if he wants to parallelize without shooting oneself in the foot. This gives the liberty to one to still use the system without "obfuscation" if there is a need for that.